crates/compiler/src/error_listener.rs

```diff
@@ -52,3 +108,127 @@ pub enum DiagnosticSeverity {
     /// to fix, but are unlikely to cause problems.
     Info,
 }
+
+pub(crate) struct LexerErrorListener {
+    pub(crate) diagnostics: RefCell<Vec<Diagnostic>>,
+    file_name: String,
+}
+
+impl LexerErrorListener {
+    pub(crate) fn new(file_name: String) -> Self {
+        Self {
+            file_name,
+            diagnostics: Default::default(),
+        }
+    }
+}
+
+impl<'input, T: Recognizer<'input>> ErrorListener<'input, T> for LexerErrorListener {
+    fn syntax_error(
+        &self,
+        _recognizer: &T,
+        _offending_symbol: Option<&<T::TF as TokenFactory<'input>>::Inner>,
+        line: isize,
+        column: isize,
+        msg: &str,
+        _error: Option<&ANTLRError>,
+    ) {
+        let line = (line - 1) as usize;
+        let column = column as usize;
+        let range = Position {
+            line,
+            character: column,
+        }..=Position {
+            line,
+            character: column + 1,
```

Comment:
why the +1?

Reply:
All things that are printed to the user are 1-based: line 1, character 1 instead of line 0, character 0.

Reply:
ohh ok